### DISCUSSION TITLE
PocketHost 0.5.3 release

### DISCUSSION THREAD

**Discussion by benallfree** [pockethost/pockethost#111] [Show and tell]
- [x] fix: incorrect instance information displaying on dashboard details in some cases
- [x] fix: more helpful error message when backup fails for nonexistent instance
- [x] chore: move version number to base package.json
- [x] refactor: logging and async helpers
- [x] chore: restore auto-cancellation
- [x] chore: rebuild with go 1.19.3 and include in bin name
- [x] fix: Disallow backups if data dir doesn't exist

**Comment by cheskoxd** [comment]
Got a quick question, which might be kinda silly, but I would like to know how do you create multiple instances (apps) of pocketbase within the same site

Im talking about the https://pockethost.io/dashboard site, there we can create a new instance of pocketbase by clicking one button
If you could explain me what is happening at the background Ill appreciate it @benallfree

**Reply by benallfree** [reply]
Hi @cheskoxd, 

All of the PockerHost code is in this repository. It uses a master copy of PocketBase to keep track of all the instances and users, and has a UI where users can log in and manage their instances. 

On the server, it launches separate instances of PocketBase for each instance defined in the UI, and shuts them down when they’re idle in order to conserve resources. 

Early on, I asked the PocketBase team if they intended to add multitenant features to PocketBase. That would make it possible to do everything in just one instance, but it was decided that it was too complicated. In hindsight, I agree with this choice because it makes individual instances much more portable if someone decides to move off the PocketHost platform and host things themselves. 

Does that answer your question?

**Reply by cheskoxd** [reply]
Yup, thanks a a lot

One last question, which hosting service are you currently using to host and manage that amount of intances running in separate subdomains?

Im asking because it seems to be so complex to handle that amount of instances, what are doing server side? 
I tried to recreate pockethost, but in the most basic way, to try to understand how pockethost is working, by using node js I was able to create a function that create a copy of the unziped pb_data file  and then run the instance as well with nodejs, so in that way I can have multiple databases,  but I was not able to run multiple instances at the same time, I was able to create multiple databases but I could only run one at a time (even using different ports for each app)

Basically my idea to achieve its a way of creating multiple instances of pocketbase and assign one (instance) per user, so each user will have their own database for my app (with a subdomain)

Why am I doing this? Because in that way each user can hold up to 10k simultaneous connections (the app will read and write a lot so, theyll need more than just a thousand of connections), rather than having only one db for all the users

Ill have a master db for all the users db subdomains so I can run an admin client to update the collections schemas or if I need to create more collections through all the instances

**Reply by cheskoxd** [reply]
@benallfree Thanks for your responses!

**Reply by benallfree** [reply]
> One last question, which hosting service are you currently using to host and manage that amount of intances running in separate subdomains?

pockethost.io currently runs on Digital Ocean on a single server. As the community grows, I upgrade the server resources. 

> Im asking because it seems to be so complex to handle that amount of instances, what are doing server side?

It shuts down instances that are no longer in use. That way, a lot of instances can share a single server. 

> I was not able to run multiple instances at the same time, I was able to create multiple databases but I could only run one at a time (even using different ports for each app)

I’m not sure why you had trouble. Multiple instances of PocketBase can definitely run at the same time. 

> Basically my idea to achieve its a way of creating multiple instances of pocketbase and assign one (instance) per user, so each user will have their own database for my app (with a subdomain)

That is essentially what pockethost does. Why not use it instead?

> Why am I doing this? Because in that way each user can hold up to 10k simultaneous connections (the app will read and write a lot so, theyll need more than just a thousand of connections), rather than having only one db for all the users

If that is the only reason, I would wait until it becomes a problem first. Optimizing this early will introduce complexity in your code and divert your energy from writing the great features that will attract users. 

There are a lot of ways pockethost could hit scaling problems, but until they actually become a problem I am not going to worry or write code to solve them. At first. PocketHost just launched a new PocketBase instance for every single instance defined. Eventually the server ran out of memory so I added code to shut down idle instances. That’s an example of a just-in-time optimization. If I had tried to do it earlier, I would have guessed the wrong approach and spent time on something that wasn’t even correct. 

So that’s my advice for you :) Stay with one instance and put all your users there until it becomes a real life problem. When it does, there will be an obvious solution that may or may not be the one you are imagining now.
